### ISSUE TITLE
NamedTuples not being documented properly

### ISSUE THREAD

**Issue by fabianp** [google-deepmind/optax#669]
Note the redundancy and all the "Alias for field number X" in classes such as optax.MultiStepsState: https://optax.readthedocs.io/en/latest/api/optimizer_wrappers.html#optax.MultiStepsState

![image](https://github.com/google-deepmind/optax/assets/277639/3c755c6e-4f3b-45b6-bb58-e6e3e9be2ead)

**Comment by amosyou** [comment]
I took a look at using [sphinx-toolbox](https://sphinx-toolbox.readthedocs.io/en/latest/extensions/more_autodoc/autonamedtuple.html) to autodoc the namedtuples, but I wasn't able to resolve the duplicate fields showing up. It seems that this is a known issue according to [here](https://github.com/sphinx-doc/sphinx/issues/8717), not too sure how else to go about fixing this.

**Comment by fabianp** [comment]
Thanks for looking into it @amosyou . Eliminating the "Alias for field ___" would already be better, even if we loose the type information IMO. Could you post a screenshot of how that would look like?

**Comment by amosyou** [comment]
<img width="831" alt="Screenshot 2024-02-03 at 12 48 46 AM" src="https://github.com/google-deepmind/optax/assets/91300605/ceda0433-9046-48cd-b21a-8e8b9ee8dc00">

**Comment by fabianp** [comment]
I think that looks good! (module some formatting issues that I thin are orthogonal to this issue). Could you submit a pull request for this?

You're on 🔥 @amosyou !

**Comment by amosyou** [comment]
revisiting this! so removing :members: from the autoclass directives isn't a viable solution since it also removes fields that are accurately documented like for LookaheadState

<img width="824" alt="Screenshot 2024-02-14 at 11 19 07 PM" src="https://github.com/google-deepmind/optax/assets/91300605/e1a9a3cf-4113-4989-9636-7c23f3a1715b">

i've tried using the autodoc-skip-member event as suggested [here](https://stackoverflow.com/questions/61572220/python-sphinx-namedtuple-documentation) but it doesn't seem to be removing the duplicate alias fields. i've also tried modifying the _monkey_patch_doc_strings in conf.py to replace lines with "Alias for field number" with empty string, also didn't work. not sure where to go from here 😅

**Comment by vroulet** [comment]
Thank you @amosyou for looking into this!
I've just tried removing :members: from the LookAheadState and the `fast_state` and `steps_since_sync` were still in the doc. One can even still have types if we are really brave:
```
class LookaheadState(NamedTuple):
  """State of the `GradientTransformation` returned by `lookahead`.

  Attributes:
    fast_state (:class:`optax.OptState`): Optimizer state of the fast optimizer.
    steps_since_sync (``jnp.ndarray``): Number of fast optimizer steps taken since
      slow and fast parameters were synchronized.
  """
  fast_state: base.OptState
  steps_since_sync: jnp.ndarray
```

<img width="818" alt="Screenshot 2024-02-15 at 08 42 34" src="https://github.com/google-deepmind/optax/assets/7056736/2e2706b9-6070-4760-8ed0-72452a8d2242">

**Comment by fabianp** [comment]
@vroulet taht solutions looks great to me!

**Comment by vroulet** [comment]
Ok, but I may have missed something, @amosyou can you try again on your side and see if that works?. Your idea of removing the :members: sounded great :). I let you see if you want to open a PR to remove the :members: and (if you are extra brave) adding the types.

**Comment by amosyou** [comment]
oh that looks good! hmm I just tried commenting out :members: again and the entire docstring for LookaheadState is gone from the docs (this has never happened before).

<img width="807" alt="Screenshot 2024-02-15 at 12 05 16 AM" src="https://github.com/google-deepmind/optax/assets/91300605/75e3d435-6aa1-4a90-b981-07719447d884">

**Comment by vroulet** [comment]
Strange. All states are missing in that screenshot (they're in pink not in blue). There may be another issue?
Here is the modification of the code on my side:
```
Flatten
~~~~~~~~
.. autofunction:: flatten

Lookahead
~~~~~~~~~~~~~~~~~
.. autofunction::  lookahead
.. autoclass::  LookaheadParams
   :members:
.. autoclass::  LookaheadState

Masked update
~~~~~~~~~~~~~~
.. autofunction::  masked
.. autoclass::  MaskedState
   :members:
```
So really just removing the line :members: below LookaheadState

**Comment by amosyou** [comment]
my bad! commenting in rst is not with # that was why they were missing 🤥

on a side note, I think to have proper rendering of the docstring we need to change Fields to Attributes (even though we're working with namedtuples here since we're using autoclass). hope that's fine?

i'll create a PR for this. thanks @vroulet!

**Comment by vroulet** [comment]
Yes, typically, MultiStepState should be reformatted with attributes instead of fields. Thanks again for looking into that!

**Comment by amosyou** [comment]
also for ApplyIfFiniteState, I noticed that all the fields are typed as Any. Did we figure out a solution for the typing jnp.array?

**Comment by vroulet** [comment]
> also for ApplyIfFiniteState, I noticed that all the fields are typed as Any. Did we figure out a solution for the typing jnp.array?

They should be `Union[jax.Array, int]` for notfinite_count, last_finite, and total_notfinite, I believe.
For inner_state it should be the usual `base.Optstate`

**Comment by vroulet** [comment]
> my bad! commenting in rst is not with # that was why they were missing 🤥
> 
> on a side note, I think to have proper rendering of the docstring we need to change Fields to Attributes (even though we're working with namedtuples here since we're using autoclass). hope that's fine?
> 
> i'll create a PR for this. thanks @vroulet!

PS: no need to add types for all attributes (it's possible but it would take forever). Simply remove the :members: in the doc. If people want to know the types, they should be able to click on [source] in the doc.
That said, if you find some issues (like the fields in MultiStepsState (which is badly formatted anyway) or like in the ApplyFiniteState you pointed out), would be super cool if you could fix them by the way. We can review that in your PR together.

**Comment by vroulet** [comment]
This was solved in #941
